generator/src/lib.rs

```diff
@@ -1705,66 +1755,70 @@ pub fn derive_setters(
+                            _ => unreachable!("Unexpected const-ness type: {}", param_ty[0]),
+                        };
+                        let mut slice_type = param_ty[1].parse::<TokenStream>().unwrap();
 
-                        let slice_param_ty_tokens;
-                        let ptr;
-                        if param_ty_string.starts_with("*const ") {
-                            let slice_type = &param_ty_string[7..];
-                            if slice_type == "c_void" {
-                                slice_param_ty_tokens = "&'a [u8]".to_string();
-                                ptr = ".as_ptr() as *const c_void";
-                            } else {
-                                slice_param_ty_tokens = "&'a [".to_string() + slice_type + "]";
-                                ptr = ".as_ptr()";
-                            }
+                        let mut ptr = if mutable {
+                            quote!(.as_mut_ptr())
                         } else {
-                            // *mut
-                            let slice_type = &param_ty_string[5..];
-                            if slice_type == "c_void" {
-                                slice_param_ty_tokens = "&'a mut [u8]".to_string();
-                                ptr = ".as_mut_ptr() as *mut c_void";
-                            } else {
-                                slice_param_ty_tokens = "&'a mut [".to_string() + slice_type + "]";
-                                ptr = ".as_mut_ptr()";
+                            quote!(.as_ptr())
+                        };
+
+                        // Interpret void array as byte array
+                        if param_ty[1] == "c_void" {
+                            let mutable = if mutable { quote!(mut) } else { quote!(const) };
+
+                            slice_type = quote!(u8).into();
+                            ptr = quote!(#ptr as *#mutable c_void);
+                        };
+
+                        let mutable_ref = if mutable { quote!(mut) } else { quote!() };
+
+                        let (slice_param_ty_tokens, set_size_stmt) = if array_size.contains("ename:") {
+                            // Should contain the same minus `ename:`-prefixed identifiers
+                            let array_size = field.c_size.as_ref().unwrap();
+                            let c_size = convert_c_expression(array_size);
+
+                            (
+                                quote!(&'a #mutable_ref [#slice_type; #c_size]),
+                                quote!(),
+                            )
+                        } else {
+                            let array_size_ident = Ident::from(array_size.to_snake_case().as_str());
+                            (
+                                quote!(&'a #mutable_ref [#slice_type]),
+                                quote!(self.inner.#array_size_ident = #param_ident_short.len() as _;),
+                            )
+                        };
```

Comment:
I don't really like hacking in this static size in `vkxml::ArrayType::Dynamic`, but this field is weird: it's definitely a pointer according to the `p` prefix and the header:
https://github.com/KhronosGroup/Vulkan-Headers/commit/d2308183f2846a1453ce7ae8c77ea7b5a7424ad2#diff-e222ae95c2b0d5082b94d6086fb1c24da18ee31384c1a39840df3b9152023ee6R11550, but the size of the data it points to is statically known. Perhaps we need a third state for this in `vkxml::Field`?

Otherwise I'd have implemented it something like this:
https://github.com/MaikKlein/ash/commit/a9b79a38abdde297d9e442e998da03ba441d87e0

Reply:
Spent some more time on this over the weekend (have an itch for perfecting "irrelevant" bits of code), and with there being no proper way to represent this `const uint8_t (* versionData)[2 * VK_UUID_SIZE]` in `vk.xml` it seems easiest to override the array type to static straight after parsing the file, or adding the "static `len` expression" heuristic to vk-parse instead.

That'd look about like this: https://github.com/Traverse-Research/ash/compare/cleanup..simplify
- Note that it includes changes to how non-const pointers in `setters` are now treated as safe references as well, and instead providing an override for "really void" things;
  - (This makes sense for things like `PipelineCreationFeedbackEXT` in the builder that are now passed as `*mut` :thinking:);
  - `ANativeWindow` should likely be void as well;
- `vkxml` really needs to add `Eq`, `Clone` (and `Copy`) to their types, but the repo hasn't been touched since 2017...
- The heuristic is wrong, and should be moved to `vk-parse` as this nested `for` is beyond unacceptable;
- Finally, `data_void` is now a proper `*const [u8; 2 * UUID_SIZE]` instead of `*const u8` in the struct :grimacing: